src/os/README.md

```diff
@@ -50,19 +50,21 @@ inspiration for several decisions in this library.
 - do not use `OS\ErrnoException` if the error condition would not be indicated
   by the `errno` variable in C. Consider adding another similar class, e.g.
   add `OS\HErrnoException` if you want to report an error exposed via `h_errno`
-- add and use Hack classes (not type aliases) for 'handle'-like parameters and
-  return values, e.g. `OS\open()` returns a `HH\Lib\FileDescriptor` instead of
-  an `int`; as well as aiding type safety, this prevents requests from interfering
-  with resources that belong to another request.
+- add and use Hack classes (not type aliases) for long-lived 'handle'-like
+  parameters and return values, e.g. `OS\open()` returns a
+  `HH\Lib\FileDescriptor` instead of an `int`; as well as aiding type safety,
+  this prevents requests from interfering with resources that belong to another
+  request.
 - Avoid `inout` parameters; return tuples instead. For example, prefer
   `function mkstemp(string $pattern): (FileDescriptor, string)` to
   `function mkstemp(inout string $in_pattern_out_path): FileDescriptor`
   - this can aid for common use by allowing string literals, rather than
     requiring otherwise-unused locals
-  - if the primary purpose of a function is to mutate its parameters, this
-    should be considered case-by-case; for example, if we were to add something
-    similar to `array_pop()`, it probably should take an `inout` container - but
-    probably shouldn't be in `OS\`
+  - if the primary purpose of a set of functions is to create, mutate and
+    destroy a C data structure, whose reference would not be held by C
+    libraries, they should be exposed as a `vec`. See [Appendix: reference
```

Comment:
> they should be exposed as a vec

Is 'they' the C structure, or the functions?

This recommendation feels pretty specific to posix_spawn_file_actions - e.g. it doesn't feel like it would apply to anything related to `sockaddr`. It shouldn't replace the previous item, but it does provide a specific exception in one case.

Perhaps `s/a C data structure/a C data structure that represents a list/` ?

Reply:
The problem is that there is not a good definition of "represents a list". Technically a `vec` can represent an arbitrary series of function calls, as long as they are setters, where "setter" is defined as a function that just changes the data structure and does not perform global side effect.

Reply:
Since there is no setter functions for `sockaddr` in the header, I would rather not include its encoding in this section.

Reply:
> The problem is that there is not a good definition of "represents a list".

It doesn't need to be 100% unambiguous; refining the definition is definitely something that can be left for later - actually following anything in these documents will still require code review, so there will be human judgement later in the process.

Yes, it's possible to represent an arbitrary series of function calls, but as a general solution, it's not fitting with the idea of merely trying to map C idioms to the corresponding Hack ones.

Reply:
> Is 'they' the C structure, or the functions?

'They' include the C pointer or handle type, the constructor, the setters, and the destructor.

Reply:
> Yes, it's possible to represent an arbitrary series of function calls, but as a general solution, it's not fitting with the idea of merely trying to map C idioms to the corresponding Hack ones.

This section describes the default encoding for setters. We can describe other special cases to map all C idioms to Hack in other sections, especially the `shape` encoding for unordered setters.

Reply:
I think our disagreement is that I consider the 'vec-like' thing to be the special case which should be an exception, rather than it being the general case with other things being exceptions :)

Reply:
In fact "they should be exposed as a `vec`" will be modified to "they should be exposed as a `vec` or `shape`." in https://github.com/hhvm/hsl/pull/175

Reply:
ok, the current PR + diff stack structure is a bit hard to review :) #175 still mentions 'long-lived' which I don't think is intentional.

I think this lines up with what you're writing and referential transparency, but to check: imagine 3 APIs:

# C++ byref

```
FooStruct myFoo;
do_foo(myFoo);
```

# C pointer-to-value

```
FooStruct myFoo;
do_foo(&myFoo);
```

# C opaque handle

```
FooStruct* myFoo = foo_init();
do_foo(myFoo);
foo_destroy(myFoo);
```

---

Assuming that all 3 only differ in terms of resource management and calling convention, these 3 approaches should be considered equivalent from the perspective of Hack API design; which approach the C library actually takes is an implementation detail. If libfoo changes between these approaches in `v(X+1)`, we should not feel any need to change the corresponding Hack APIs.

Is this view compatible with your plans here?